brave/brave-core
# Sync warn keyring locked

Resolves https://github.com/brave/brave-browser/issues/24121

This PR does 4 things:

1. Improvement of `TimeLimitedWords::GenerateForDate` / `TimeLimitedWords::Parse`. Now it returns base::expected result. This prevents the crash mentioned in issue and increases readability.
2. brave://sync-internals page is expanded with `Brave Sync` section and boolean fields `Passphrase is set`, `Passphrase decryption failed`, `OS encryption available` to make easier diagnostic of similar cases.
3. When Brave Sync has the passphrase set in preferences, but by any reason failed to decrypt it, brave://settings/braveSync/setup page displays `Browser could not decrypt the passphrase, Brave Sync is not running. Please check that browser has access to Keychain` message . The commit should be discussed, maybe it is not so necessary.
4. Adds the informer about Sync is not running for the case when decrypt seed failed.

**Screenshots:**

![image](https://user-images.githubusercontent.com/24739341/184966386-1fe88406-0386-4007-baa0-97bde8db612f.png)
![Screenshot from 2022-07-20 20-21-15](https://user-images.githubusercontent.com/24739341/180044158-944d4427-039e-4bae-ad8e-1efb7dbfe762.png)
![image](https://user-images.githubusercontent.com/24739341/180766115-dcd25d65-0d18-4c2c-8b5c-4dff766efe39.png)




## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
Test plan

1. Take clean profile
2. Launch browser
3. Open brave://sync-internals page. Expected to have 


| Brave Sync | |
| --- | --- |
| Passphrase is set | false |
| OS encryption available | true |

5. Create a new sync chain. Open brave://sync-internals page. Expected to have 

| Brave Sync | |
| --- | --- |
| Passphrase is set | true |
| OS encryption available | true |

6. Exit browser
7. Lock the keychain
8. Launch browser  and don't grant the keychain access. 
9. Expected to see informer `Sync service could not be started, Sync is not running`
10. Click `Check details` button - expected to open brave://settings/braveSync/setup page
11. Expected to see `Browser could not decrypt the passphrase, Brave Sync is not running. Please check that browser has access to secure OS passwords store. If you are using macOS, this is Keychain app. If you are using Ubuntu, this is Passwords and Keys app. Also don't use copied profile from other machines, this is not supported for security reasons.` . 
12. Open  brave://sync-internals page,  expected to see 

| Brave Sync | |
| --- | --- |
| Passphrase is set | true |
| Passphrase decryption failed | true |
| OS encryption available | false |

13. Exit browser, unlock the keychain, launch browser grant sync chain access.
14. Expected to see on brave://settings/braveSync/setup the regular settings, devices and data types.

Informer logic test:
1. Lock the keychain
2. Launch browser  and don't grant the keychain access. 
3. Expected to see informer `Sync service could not be started, Sync is not running`
4. Press `Don't show again`
5. Restart browser, don't grant keychain access
6. Expected not to see informer `Sync service could not be started, Sync is not running`
7. Restart browser, grant keychain access
8. Leave sync chain, create new sync chain
9. Restart browser, don't grant keychain access
10. Expected to see informer `Sync service could not be started, Sync is not running`
